// hero-slider/dist/components/SideNav/index.js
import React__default from 'react';
import SideNavModuleCss from './index.module.css.js';
import { useController } from '../../modules/Controller.js';
import { useManager } from '../../modules/Manager.js';
import { composeCssClasses } from '../../utils/composeCssClasses.js';

function SideNav({ color, activeColor, left, right, isPositionedRight = true, position = {
    bottom: undefined,
    top: '50%',
    left: !isPositionedRight ? left || '1rem' : undefined,
    right: isPositionedRight ? right || '1rem' : undefined,
    transform: 'translateY(-50%)'
} }) {
    const { state: { slides, totalSlides } } = useManager();
    const { state: { activeSlide }, changeSlide } = useController();
    function renderButtons() {
        if (!totalSlides)
            return [];
        const onClickHandler = (slideNumber) => {
            if (slideNumber !== activeSlide)
                changeSlide(slideNumber);
        };
        return Array.from(slides.values()).map(({ number }) => {
            return (
            // TODO: Deal with the disabled linting later:
            // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions, jsx-a11y/click-events-have-key-events
            React__default.createElement("li", { key: number, className: composeCssClasses('hero-slider-side-nav-button', SideNavModuleCss.Button, {
                    className: SideNavModuleCss.Active,
                    useIf: activeSlide === number
                }), onClick: () => onClickHandler(number), style: {
                    justifyContent: isPositionedRight ? 'flex-end' : 'flex-start'
                } },
                React__default.createElement("span", { className: composeCssClasses('hero-slider-side-nav-button-line', SideNavModuleCss.Line) }),
                React__default.createElement("span", { className: composeCssClasses('hero-slider-side-nav-button-number', SideNavModuleCss.Number) }, number.toLocaleString())));
        });
    }
    /**
     * CSS variables for the transitions.
     */
    const CSSVariables = {
        '--nav-color': color,
        '--nav-active-color': activeColor
    };
    return (React__default.createElement("ul", { className: composeCssClasses('hero-slider-side-nav-wrapper', SideNavModuleCss.Wrapper), style: Object.assign(Object.assign({}, position), CSSVariables) }, renderButtons()));
}
SideNav.displayName = 'hero-slider/nav';

export { SideNav };
