// hero-slider/dist/components/AutoplayButton/index.js
import React__default from 'react';
import AutoplayButtonModuleCss from './index.module.css.js';
import { useAutoplay } from '../../modules/Autoplay.js';
import { IntervalState } from '../../modules/IntervalTimer.js';
import { composeCssClasses } from '../../utils/composeCssClasses.js';

var ButtonType;
(function (ButtonType) {
    ButtonType["PLAY"] = "play";
    ButtonType["PAUSE"] = "pause";
})(ButtonType || (ButtonType = {}));
class AutoplaySvg {
    static getPath(buttonType) {
        if (buttonType === ButtonType.PAUSE)
            return AutoplaySvg.pausePath;
        else
            return AutoplaySvg.playPath;
    }
}
AutoplaySvg.playPath = 'M 12,26 18.5,22 18.5,14 12,10 z M 18.5,22 25,18 25,18 18.5,14 z';
AutoplaySvg.pausePath = 'M 12,26 16,26 16,10 12,10 z M 21,26 25,26 25,10 21,10 z';
function AutoplayButton(props) {
    const { className, style, position = {
        bottom: '0',
        left: '0'
    } } = props;
    const { state: { isPausedByUser }, autoplayState, resume, pause } = useAutoplay();
    const [buttonType, setButtonType] = React__default.useState(autoplayState !== IntervalState.IDLE ? ButtonType.PAUSE : ButtonType.PLAY);
    const onClickHandler = (event) => {
        event.stopPropagation();
        if (isPausedByUser)
            resume();
        else
            pause();
    };
    React__default.useEffect(() => {
        if (isPausedByUser && autoplayState === IntervalState.IDLE)
            setButtonType(ButtonType.PLAY);
        else if (isPausedByUser && autoplayState === IntervalState.PAUSED)
            setButtonType(ButtonType.PLAY);
        else
            setButtonType(ButtonType.PAUSE);
    }, [isPausedByUser, autoplayState]);
    return (React__default.createElement("button", { className: composeCssClasses('hero-slider-autoplay-button', AutoplayButtonModuleCss.Button, className), onClick: onClickHandler, style: Object.assign(Object.assign({}, position), style) },
        React__default.createElement("svg", { fill: "currentColor", height: "100%", width: "100%", viewBox: "0 0 36 36" },
            React__default.createElement("path", { d: AutoplaySvg.getPath(buttonType) }))));
}
AutoplayButton.displayName =
    'hero-slider/autoplay-button';

export { AutoplayButton };
